WIP: Add a new example project for showcasing the presentation styles and options

nataliq opened the pull request:
## In this PR..
I created a new example project with the main goal of adding UI tests to it so that we can test our presentations. It focuses mainly on the styles and options - presenting dummy screens with different combinations of style + options.

It's definitely not covering all combinations since not all of them make sense or are interesting for testing. It might miss some interesting cases though - my idea is to have this project as a starting point and add more cases as we discover them.

## Potential areas for further improvements
- filter the presentation options that make sense for each presentation style
- allow multiple options selection
- cover `DualNavigationControllersSplitDelegate` and `MasterDetailSelection` functionality
- have presentations completing with Signals (currently covering only Futures)

_Do you think I should address any of those before merging?_

## Out of the scope
I intentionally didn't include any extensive tests for our Alerts, I was thinking to add a separate project for that.

_Does this make sense?_

## TODO
- [ ] Check that the projects builds properly with the build script that integrates with CocoaPods
- [ ] Add a build step to execute the UI tests

nataliq commented:
@iZettle/ios-devs can someone take a look?

fennecos commented:
What do think about showing the presentation options in an action sheet instead?
Feels a bit strange to show a full screen view controller

fennecos commented on Examples/StylesAndOptions/Example/AppFlow.swift:
```diff
@@ -0,0 +1,77 @@
+            .present(ChooseStyle(), options: [.defaults, .showInMaster])
+            .onValue { style, preferredPresenter, alertToPresent in
+
+                containerController
```
the result of present should probably be added to the bag.
Otherwise if the bag gets disposed, the presentation won't get disposed.

But maybe we want to flatMap instead?

fennecos commented on Examples/StylesAndOptions/Example/ChooseOptions.swift:
```diff
@@ -0,0 +1,59 @@
+}
+
+struct ChoosePresentationOptions {
+    private let createDataSource: () -> DataSource<NamedPresentationOptions>
```
why is this a closure? seems like it could be a property

mansbernhardt commented on Examples/StylesAndOptions/Example/AppFlow.swift:
```diff
@@ -0,0 +1,77 @@
+            .onValue { style, preferredPresenter, alertToPresent in
+
+                containerController
+                    .present(Presentation(ChoosePresentationOptions(), style: .modal, options: [.defaults, .showInMaster], configure: installDismiss)).future
```
Do we need to pass `.showInMaster` here as we present modally, hence default options should do?

mansbernhardt commented on Examples/StylesAndOptions/Example/ChooseOptions.swift:
```diff
@@ -0,0 +1,59 @@
+import Flow
+import Presentation
+
+struct NamedPresentationOptions {
```
This can potentially just be a type alias to (name: String, options: PresentationOptions)

nataliq replied:
I find the struct more convenient since it allows me to use a helper for creating data sources for types conforming to CustomStringConvertible

mansbernhardt commented on Examples/StylesAndOptions/Example/ChooseStyle.swift:
```diff
@@ -0,0 +1,64 @@
+}
+
+extension ChooseStyle: Presentable {
+    typealias ChooseStyleResult = (PresentationStyle, preferredPresenter: UIViewController?, alertToPresent: Alert<()>?)
```
Trying to follow we can't just signal the style here? 
- Not sure why we need `preferredPresenter`? What would happen if we pass nil for "embed"?
- The `alertToPresent` seems to be used by "sheet" and uses a constant value, so this check could be moved to the other end?

nataliq commented:
I finally picked this up and addressed the PR feedback. I added a short readme too. The leaking presentations from the AppFlow should be fixed now but you can take another look at the presentations chain:
```swift
bag += styleWithContext.onValueDisposePrevious { style, preferredPresenter, alertToPresent in
    return chooseOptions().onValue { options in
        bag += showDismissablePresentationOrAlert(style, options, preferredPresenter, alertToPresent).disposable
    }.disposable
}
```
I rebased to get the changes from 1.1 so the new commits start with `Make sure presentations are not leaking when bags are disposed`. 

Lastly, I see very little value with making this look nicer or do something otherwise useful, so I suggest if you're ok with the changes to get this merged and if someone wants to iterate on it - he/she is welcome to do so :) 

@fennecos @mansbernhardt

nataliq commented:
Not sure what was wrong with the CI, I couldn't restart the stale build. Triggered a new one with a small update, can I get another +1?
